RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,210 @@
+
+It is a quality of implementation how equality of offsets and scalings are checked.
+
+`Empty` units and `undefined` units are always compatible with each other, themselves and any `well-formed` units.
+
+## Unit inference
+The following rules are used to perform unit inference of any expression.
+They describe the unit meta-expression of a given expression, where it makes sense.
+They also provide constraints that arise from different Modelica constructs.
+When it is specified that _`e1` must be unit compatible with `e2`_, we mean that the corresponding constraint should be added to the inference algorithm.
+
+After collecting all constraints, we can run a pre-processing phase that evaluates both sides of the constraints.
+Any constraints of the form _`u` must be compatible with empty unit_, or _`u` must be compatible with undefined unit_, are trivially satisfied and shall be discarded when running the inference algorithm.
+
+In this section, the unit meta-expression of expression `e` is called `e.unit`.
+
+### Variables
+Consider the expression `e` of the form `var`, where `var` is a Modelica variable.
+* If `var` doesn't have a declared unit attribute, and has constant component variablity, `e.unit` is equal to `empty` unit.
+* Otherwise, `e.unit` is equal to `var.unit`.
+
+If `var` has a declared unit attribute, `var.unit` must be compatible with it.
+
+### Literals
+Literals have `empty` unit.
+
+### Binary operations
+Consider an expression `e` of the form `e1 op e2`.
+
+#### Additive operators
+* `e1.unit` and `e2.unit` must be unit compatible.
+* `e.unit` must be compatible with `e1.unit`.
+* `e.unit` must be compatible with `e2.unit`.
+
+The three constraints ensure unit propagation even when one of the operand units is `empty` or `undefined`
+
+#### Multiplication
+`e.unit` is equal to `e1.unit * e2.unit`.
+
+#### Division
+`e.unit` is equal to `e1.unit * (e2.unit ^ -1)`.
+
+#### Relational operators
+The unit of `e1` and the unit of `e2` must be unit compatible.
+
+#### Power operator
+* If `e2` is a constant or evaluated Integer, `e.unit` is equal to `e1.unit ^ e2`. (Note that unit-checking does not trigger evaluation.)
+* Otherwise, `e1.unit` must be compatible with `"1"` and `e.unit` is equal to `"1"`.
+
+(If the exponent was not evaluated or is a Real, the unit of the base should be `"1"`.)
+
+(It is up for discussion, when `e2` is Real, whether `e2.unit` should be compatible with `"1"`.)
+
+### Function calls
+* For each expression passed as argument, its unit must be compatible with the declared unit of the corresponding function input.
+* If the output of the function has a declared unit, the unit of the expression is the declared unit.
+* The unit of the expression is `undefined` otherwise.
+
+### Operator der
+The unit of the expression `der(e)` is `der(e.unit)`.
+
+### If expressions
+Consider the expression `e` of the form `if cond then e1 else e2 end if;`
+* `e1.unit` and `e2.unit` must be unit compatible.
```

Comment:
I think this rule deserves a rationale after the list of rules, because it looks a bit redundant in view of the other two… assuming it actually isn't redundant?!